packages/core/src/processors/runner.ts

```diff
@@ -0,0 +1,291 @@
+            let state = processorStates.get(processor.name);
+            if (!state) {
+              state = new ProcessorState(processor.name);
+              processorStates.set(processor.name, state);
+            }
+
+            // Add the current chunk to accumulated text
+            state.addChunk(processedChunk);
+
+            const result = await processor.processOutputStream({
+              chunk: processedChunk,
+              allChunks: state.allChunks,
+              state: state.getState(),
+              abort: (reason?: string) => {
+                throw new TripWire(reason || `Chunk blocked by ${processor.name}`);
+              },
+            });
+
+            processedChunk = result.chunk;
+            shouldEmit = shouldEmit && result.shouldEmit;
+          }
+        } catch (error) {
+          if (error instanceof TripWire) {
+            return { chunk: processedChunk, blocked: true, reason: error.message, shouldEmit: false };
+          }
+          // Log error but continue with original chunk
+          this.logger.error(`[Agent:${this.agentName}] - Output processor ${processor.name} failed:`, error);
+        }
+      }
+
+      return { chunk: processedChunk, blocked: false, shouldEmit };
+    } catch (error) {
+      this.logger.error(`[Agent:${this.agentName}] - Chunk processing failed:`, error);
+      return { chunk, blocked: false, shouldEmit: true };
+    }
+  }
+
+  async runOutputProcessorsForStream(
+    streamResult: StreamObjectResult<any, any, any> | StreamTextResult<any, any>,
+  ): Promise<ReadableStream<any>> {
+    return new ReadableStream({
+      start: controller => {
+        const reader = streamResult.fullStream.getReader();
+        const processorStates = new Map<string, ProcessorState>();
+
+        const processChunk = async () => {
+          try {
+            const { done, value } = await reader.read();
+
+            if (done) {
+              controller.close();
+              return;
+            }
+
+            // Process all chunks through output processors
+            const {
+              chunk: processedChunk,
+              blocked,
+              reason,
+              shouldEmit,
+            } = await this.processChunk(value, processorStates);
+
+            if (blocked) {
+              // Log that chunk was blocked
+              void this.logger.debug(`[Agent:${this.agentName}] - Chunk blocked by output processor`, {
+                reason,
+                originalChunk: value,
+              });
+
+              // Send tripwire chunk and close stream for abort
+              controller.enqueue({
+                type: 'tripwire',
+                tripwireReason: reason || 'Output processor blocked content',
+              });
+              controller.close();
+            } else if (shouldEmit) {
+              // Send processed chunk only if shouldEmit is true
+              controller.enqueue(processedChunk);
```

Comment:
can't we remove shouldEmit as a concept and only enqueue it if they return a part? null/false/undefined would just mean we skip it

Reply:
was also thinking about doing this but didn't get that far